mergin/client.py

```diff
@@ -365,25 +365,31 @@ def apply_pull_changes(self, changes, temp_dir):
                 # special care is needed for geodiff files
                 if self.is_versioned_file(path) and k == 'updated':
                     if path in modified:
+                        f_server_backup = self.fpath(f'{path}-server_backup', temp_dir)
                         server_diff = self.fpath(f'{path}-server_diff', temp_dir)  # single origin diff from 'diffs' for use in rebase
                         rebased_diff = self.fpath(f'{path}-rebased', temp_dir)
-                        patchedfile = self.fpath(f'{path}-patched', temp_dir)  # patched server version with local changes
-                        changeset = self.fpath(f'{path}-local_diff', temp_dir)  # final changeset to be potentially committed
+                        shutil.copy(src, f_server_backup)  # temporary backup of file pulled from server for test and recovery
                         try:
                             self.geodiff.create_changeset(basefile, src, server_diff)
                             self.geodiff.create_rebased_changeset(basefile, dest, server_diff, rebased_diff)
-                            self.geodiff.apply_changeset(src, patchedfile, rebased_diff)
-                            self.geodiff.create_changeset(src, patchedfile, changeset)
-                            shutil.copy(src, basefile)
-                            shutil.copy(patchedfile, dest)
+                            # update file with rebased_diff to contain both server and local changes
+                            self.geodiff.apply_changeset(src, rebased_diff)
+                            # try to create final changeset to be potentially committed in push
+                            changeset = self.fpath(f'{path}-local_diff', temp_dir)
+                            self.geodiff.create_changeset(f_server_backup, src, changeset)
+                            # we are happy with rebase, prepare 'live' versions of files
+                            shutil.copy(f_server_backup, basefile)
+                            shutil.copy(src, dest)
```

Comment:
As mentioned above - we should try to avoid overwriting of files, especially the ones like 'dest' which may be open in QGIS.